### PR TITLE
Add option to build/ install with cmake

### PR DESCRIPTION
**Pull request by fferflo**
Adds option to build examples and install with cmake.

After installing with cmake, this project can be linked in another cmake project via

```
find_package(osmpbfreader REQUIRED)
target_link_library(YOURTARGET PUBLIC osmpbfreader::osmpbfreader)
```

without having to manually copy or link the header file.

### CODE REVIEW

**Comment by benoit-bst** [comment]
Hi @fferflo,
I tried to use your **cmake** into our main project with a simple `add_subdirectory(osmpbfreader)`, and an error occured.
```
CMake Error at third_party/osmpbfreader/CMakeLists.txt:12 (find_package):
  By not providing "Findosmpbf.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "osmpbf", but
  CMake did not find one.

  Could not find a package configuration file provided by "osmpbf" with any
  of the following names:

    osmpbfConfig.cmake
    osmpbf-config.cmake

  Add the installation prefix of "osmpbf" to CMAKE_PREFIX_PATH or set
  "osmpbf_DIR" to a directory containing one of the above files.  If "osmpbf"
  provides a separate development package or SDK, be sure it has been
  installed.

```
In standalone mode, there is no problem with `cmake . && make`

```
~/dev/osmpbfreader/build(e747c93) » cmake ..
-- The CXX compiler identification is GNU 7.5.0
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found osmpbf: /usr/include
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Found Protobuf: /usr/lib/x86_64-linux-gnu/libprotobuf.so;-lpthread (found version "3.0.0")
-- Configuring done
-- Generating done
-- Build files have been written to: /home/bbrisset/dev/navitia/source/third_party/osmpbfreader/build

~/dev/osmpbfreader/build(e747c93) » make
Scanning dependencies of target example_counter
[ 25%] Building CXX object CMakeFiles/example_counter.dir/example_counter.cc.o
[ 50%] Linking CXX executable bin/example_counter
[ 50%] Built target example_counter
Scanning dependencies of target example_routing
[ 75%] Building CXX object CMakeFiles/example_routing.dir/example_routing.cc.o
[100%] Linking CXX executable bin/example_routing
```

Even if we have no need to that, it seems to be pretty nice to add it for the simplicity and modernity. Thanks for the effort :+1:

**Comment by fferflo** [comment]
Hey @benoit-bst,

the find_package call is only necessary when installing libosmpbfreader into a different directory and linking your project to that directory. The standard workflow with cmake is:

1. Build and install libosmpbfreader either to global directories or to a custom directory by setting CMAKE_INSTALL_PREFIX
2. Add the install directory of libosmpbfreader to CMAKE_PREFIX_PATH, for example by calling `export CMAKE_PREFIX_PATH=$CMAKE_PREFIX_PATH:your_installation_path` before calling cmake on your project (or skip this if you installed to global directories)
3. Call `find_package(osmpbfreader REQUIRED)` which looks for libosmpbfreader in CMAKE_PREFIX_PATH (or globally) and creates the target `osmpbfreader::osmpbfreader`
4. Link your target to libosmpbfreader with `target_link_library(YOURTARGET PUBLIC osmpbfreader::osmpbfreader)`

With these steps your folders of libosmpbfreader and your custom project don't intermingle, and if you want to use libosmpbfreader in multiple projects you only have to download and install it once.

If you prefer the _quick-and-dirty_ approach via `add_subdirectory`, you don't have to call find_package since your project already knows where to find libosmpbfreader. I haven't tried, but in that case you should be able just link libosmpbfreader via

```target_link_library(YOURTARGET PUBLIC osmpbfreader)```

without the namespace prefix `osmpbfreader::` which is only added when calling find_package.

Glad to help!

**Comment by benoit-bst** [comment]
Thanks for the explanation.
My Cmake skills was a bit dusty :older_man:

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  INTERFACE
+    ${osmpbf_INCLUDE_DIRS}
+    ${Protobuf_INCLUDE_DIRS}
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}>
+    $<INSTALL_INTERFACE:include>
+)
+target_compile_features(osmpbfreader INTERFACE cxx_std_11)
+target_link_libraries(osmpbfreader INTERFACE
+  ${Protobuf_LIBRARIES}
+  z
```
Comment (benoit-bst) [inline CMakeLists.txt:25]: `z` is not a library ?

**Review by benoit-bst — APPROVED** [review]
@xlqian @nonifier I tested with our Navitia project and there is no impact on it.
It seems to be a good way to modernize this sub part in the future !

**Review by xlqian — APPROVED** [review]

**Pull request merged**